infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/infrastructure/msgHandlers/ReplicationHandlerMethods.java

```diff
@@ -0,0 +1,192 @@
+                .forEach(method -> handler.registerMethod(caller, server, method));
+        return handler;
+    }
+
+    private void registerMethod(@Nonnull final MethodHandles.Lookup caller,
+                                @Nonnull final LogReplicationAbstractServer server,
+                                @Nonnull final Method method) {
+        final LogReplicationMsgHandler annotation = method.getAnnotation(LogReplicationMsgHandler.class);
+
+        // check if the method is already registered
+        if((!annotation.requestType().equals(RequestPayloadMsg.PayloadCase.NONE) && requestHandlerMap.containsKey(annotation.requestType())) ||
+                (!annotation.responseType().equals(ResponsePayloadMsg.PayloadCase.NONE) &&
+                responseHandlerMap.containsKey(annotation.responseType())))  {
+            throw new IllegalStateException("HandlerMethod for " + annotation.requestType() + " already registered!");
+        }
+
+        try {
+            HandlerMethod h;
+            if (Modifier.isStatic(method.getModifiers())) {
+                MethodHandle mh = caller.unreflect(method);
+                h = (HandlerMethod) LambdaMetafactory.metafactory(caller,
+                        "handle", MethodType.methodType(HandlerMethod.class),
+                        mh.type(), mh, mh.type()).getTarget().invokeExact();
+            } else {
+                // Instance method, so we need to capture the type.
+                MethodType mt = MethodType.methodType(method.getReturnType(), method.getParameterTypes());
+                MethodHandle mh = caller.findVirtual(server.getClass(), method.getName(), mt);
+
+                // Note: mtt captures the return type as well as the argument types of the annotated handler
+                // method. These should be consistent with the handle method described by the HandlerMethod
+                // functional interface.
+                MethodType mtt = mh.type().dropParameterTypes(0, 1);
+                h = (HandlerMethod) LambdaMetafactory.metafactory(caller, "handle",
+                        MethodType.methodType(HandlerMethod.class, server.getClass()),
+                        mtt, mh, mtt).getTarget().bindTo(server).invoke();
+            }
+
+            String payloadSting = annotation.requestType().equals(RequestPayloadMsg.PayloadCase.NONE) ?
+                    annotation.responseType().toString() : annotation.requestType().toString();
+
+            // Install pre-conditions on handler
+            final HandlerMethod handler = generateConditionalHandler(payloadSting, h);
+            // Install the handler in the map
+            if (!annotation.requestType().equals(RequestPayloadMsg.PayloadCase.NONE)) {
+                requestHandlerMap.put(annotation.requestType(), handler);
+            } else {
+                responseHandlerMap.put(annotation.responseType(), handler);
+            }
```

Comment:
Looks like we put all the handler methods into `responseHandlerMap` ()? it's a bit surprising that the CICD could still pass? or I missed something here?

Reply:
I think because of the default NONE type.. it didn't throw NPE here...

Reply:
We put it int 2 different maps: requestHandlerMap has request handler methods and responseHandlerMap has response handler methods.

Sorry I did not completely get the CICD part...what/why do you expect to see failing?

Reply:
sorry I mean, in the if check, we should check is the annotation a "request" or a "response", but here we actually only check is the request type is NONE or not - which means responseHandlerMap (in else branch) takes all the response, and all the request handlers whose type is not NONE

Reply:
So I was thinking some handler methods were not even added, then why we could still make the system work? (we are passing all CICD)

Reply:
`registerMethod()` is called for every handle* method in LogReplicationServer. And the handle* method is either of request or response type.
And in this version of the PR, the responseType would be the default NONE value for a request handler.
And vice versa for responseHandler.

So checking if requestType != NONE would put all the request-handler methods in requestHandlerMap and the others (which are only the response handlers) would go to responseHandlerMap

But this condition is changing to check the type of annotation in the next push where we have 2 annotations.

Reply:
Oh okay I see the point, sorry seems like I was in a kind of lost state :) but anyway it has changed to another approach